untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,493 @@
+Indeed, we can define a combinator that would behave exactly like `if ... then ... else ...` construct: $ifThenElse = \lambda c\;t\;f. c\;t\;f.$
+Since Church boolean values themselves encode the behavior, this function doesn't do much.
+It's nice to see we can have familiar branching constructs almost right away.
+
+Now that we have booleans, we can define operations on booleans, i.e. a boolean algebra.
+We can define all the usual boolean functions in terms of Church booleans.
+
+For example $and = \lambda x\;y. x\;y\;fls.$
+The $and$ function must return $tru$ if and only if both its arguments are $tru.$
+Let us make sure it does what we expect it to.
+
+$and\;tru\;tru = (\lambda x\;y. x\;y\;fls)\;(\lambda x\; y. x)\;(\lambda x\; y. x)$
+
+We will rename the bound variables to avoid ambiguity:
+
+$(\lambda x\;y. x\;y\;fls)\;(\lambda a\; b. a)\;(\lambda c\; d. c)$
+
+Now we can reduce the expression.
+Let's use call by value strategy here.
+The arrow $\to$ represents one evaluation step:
+
+$(\lambda x\;y. x\;y\;fls)\;(\lambda a\; b. a)\;(\lambda c\; d. c)$
+$\to (\lambda y. (\lambda a\; b. a)\;y\;fls)\;(\lambda c\; d. c)$
+$\to (\lambda a\; b. a)\;(\lambda c\; d. c)\;fls$
+$\to (\lambda b. (\lambda c\; d. c))\;fls$
+$\to (\lambda c\; d. c) = tru.$
+
+So far so good.
+Now let's consider the case of $and\;fls\;tru:$
+
+$and\;fls\;tru$
+$= (\lambda x\;y. x\;y\;fls)\;(\lambda a\; b. b)\;(\lambda c\; d. c)$
+$\to (\lambda y. (\lambda a\; b. b)\;y\;fls)\;(\lambda c\; d. c)$
+$\to (\lambda a\; b. b)\;(\lambda c\; d. c)\;fls$
+$\to (\lambda b. b)\;fls$
+$\to fls.$
+
+That also works.
+The remaining two cases are left as an exercise for the reader, but it should be rather obvious that the combinator works as intended.
+Indeed, since $tru$ essentially chooses its first argument and $fls$ its second, $x\;y\;fls$ essentially means "if $x$ is $tru$, the result is $y$, otherwise $fls$".
+
+Defining other boolean combinators is reasonably straightforward by analogy.
+
+### Church pairs
+
+Since we have booleans now, we can also encode pairs (and by extension, tuples as trees or right combs) as functions, returning one value when given $tru$ as an argument, and another when given $fls$.
+
+First of all, let us define a pair constructor combinator:
+
+$pair = \lambda f\;s\;b. b\;f\;s.$
+
+It looks suspiciously similar to our $ifThenElse$ combinator, the only difference being that condition is now the last argument.
+
+Now we can define two combinators for deconstructing a pair:
+
+$fst = \lambda p. p\;tru,$\
+$snd = \lambda p. p\;fls.$
+
+It's reasonably straightforward to show that these definitions work as intended.
+Consider, for instance,
+
+$fst\;(pair\;x\;y)$
+$= (\lambda p. p\;tru)\;((\lambda f\;s\;b. b\;f\;s)\;x\;y)$
+$\Rightarrow (\lambda p. p\;tru)\;(\lambda b. b\;x\;y)$
+$\to (\lambda b. b\;x\;y)\;tru$
+$\to tru\;x\;y$
+$= (\lambda a\; b. a)\;x\;y$
+$\Rightarrow x.$
+
+Here, double arrow $\Rightarrow$ signifies more than one step of beta-reduction.
```

Comment:
multiple steps of beta-reduction?

Reply:
Technically, "multiple" is ill-defined, as it can depending on the context mean "one or more" or "more than one", but in this case it doesn't really matter. Changed.